Sources/AsyncKit/ConnectionPool.swift

```diff
@@ -138,43 +163,56 @@ public final class ConnectionPool<Source> where Source: ConnectionPoolSource  {
     ///
     /// The connection returned by this method should be released when you are finished using it.
     ///
-    ///     let conn = try pool.requestConnection().wait()
+    ///     let conn = try pool.requestConnection(on: ...).wait()
     ///     defer { pool.releaseConnection(conn) }
     ///     // use the connection
     ///
     /// See `withConnection(_:)` for a callback-based method that automatically releases the connection.
     ///
+    /// - parameters:
+    ///     - on: Preferred event loop for the new connection.
     /// - returns: A future containing the requested connection.
-    public func requestConnection() -> EventLoopFuture<Source.Connection> {
-        guard !self.isClosed else {
-            return self.source.eventLoop.makeFailedFuture(ConnectionPoolError.closed)
+    public func requestConnection(
+        on eventLoop: ConnectionPoolEventLoop = .any
+    ) -> EventLoopFuture<Source.Connection> {
+        // synchronize access to available / active connection checks
+        self.lock.lock()
```

Comment:
I like this coarse locking way better; should help prevent any race conditions e.g. between accesses to `didShutdown` and the other fields :-)